test/TEST-74-AUX-UTILS/test.sh

```diff
@@ -4,6 +4,7 @@ set -e
 
 TEST_DESCRIPTION="Tests for auxiliary utilities"
 NSPAWN_ARGUMENTS="--private-network"
+TEST_NO_NSPAWN=1
```

Comment:
I don't think this is necessary, but please confitionalize the newly added test case.

Reply:
Please see the commit message - we already switched this test in mkosi to be qemu-only via https://github.com/bluca/systemd/commit/f4faac20730cbb339ae05ed6e20da687a2868e76
I do not think this behaviour should differ between mkosi and scripts. Either it's a VM in both, or it's a container in both.

Reply:
Did you see the patch description? If we do that in the one runner, I think we want to also do it in the other.

Reply:
Then, please just control the behavior only for the runner, but do not skip running on nspawn on centos cis. Iirc, creating a dot file in TEST-74 directory controls the behavior for the runner, and that does not affect centos cis.

Reply:
That still misses the point - having different behavior between different runners means bugs get introduced and things break if you only run mkosi. That doesn't make sense. If you want test-74 to run in nspawn, please send a follow-up to fix all test cases so they can run in the nspawn backend, and then remove the configuration for both runners

Reply:
(I'm not sure why CentOS CI is compared with mkosi given that mkosi can't run things under ASan/UBSan properly. In terms of covering as many things as possible with ASan/UBSan it makes sense to run the tests in as many environments as possible there (with coverage reports of course to see what's covered instead of guessing))

@mrc0mmand what do you think?

> It doesn't make sense to have the runners behave differently, especially as the script one is on the way out.

Why would it be on the way out? Have all the issues with ASan/UBsan, coredumps and coverage reports been addressed in mkosi?

Reply:
No no no, NAK! Please do not kill tests on CentOS CIs needlessly. CentOS CIs run all integration tests both on nspawn and qemu as possible as they can. That improves coverage, as some tools behaves differently on qemu and nspawn. If nspawn test is disabled, that means many tests are tested only on qemu. That's horrible.

In principle, I'd like to make other test runners should run the integration tests both on qemu and nspawn, rather than KILLING one of them! But, I do not request that at least in this PR, as that needs more workload, hence GitHub actions, semaphore, ubuntu CIs will easily timeout.

> That still misses the point - having different behavior between different runners means bugs get introduced and things break if you only run mkosi. That doesn't make sense.

Cannot follow. Most tools should work in both qemu and nspawn, at least gracefully. Of course, there is always exceptions, e.g. udev cannot run on nspawn. We must test tools on both environment. If we find any issue when running on nspawn, then that's really nice! We need to fix the relevant code. Happy to know that.

> If you want test-74 to run in nspawn, please send a follow-up to fix all test cases so they can run in the nspawn backend, and then remove the configuration for both runners

IIUC, the test can run on nspawn. The commit by Daan means that the coverage level on qemu is higher than nspawn, so we pick qemu. That's all. And that does not mean running on nspawn is meaningless.

Reply:
Hm, there is TEST_PREFER_NSPAWN, but there is seems TEST_PREFER_QEMU. Though, I am not familiar the option.

Reply:
> No no no, NAK! Please do not kill tests on CentOS CIs needlessly. CentOS CIs run all integration tests both on nspawn and qemu as possible as they can. That improves coverage, as some tools behaves differently on qemu and nspawn. If nspawn test is disabled, that means many tests are tested only on qemu. That's horrible.

They don't, it's one or the other. At some point when we have time to port it over, the centos CI will move to mkosi too, so it won't be killed, don't worry. The legacy test framework will though, as it's just a burden at this point.

> In principle, I'd like to make other test runners should run the integration tests both on qemu and nspawn, rather than KILLING one of them! But, I do not request that at least in this PR, as that needs more workload, hence GitHub actions, semaphore, ubuntu CIs will easily timeout.

Nah, that would be a waste of time. It's fine to use nspawn where possible as it's faster, but for some things VM isolation is needed.

> IIUC, the test can run on nspawn. The commit by Daan means that the coverage level on qemu is higher than nspawn, so we pick qemu. That's all. And that does not mean running on nspawn is meaningless.

It is meaningless, as half the things are silently skipped. Nobody runs the legacy shell tests locally anymore, so things just go unnoticed. They need to be in sync, and it's a bug that they are not.

So once again, if you'd like TEST-74 to run in nspawn, please send a follow-up PR later to move the tests that require VM isolation out of it, and into another test that is marked for VM, and then it can be moved back to not requiring a VM. The current setup is broken and must be fixed.

Reply:
IIUC, still CentOS CIs have the highest coverage level over other CIs. And, even with the high coverage, CentOS CIs are quite stable and well maintained, unlike ubuntu CIs and semaphore CIs. The GitHub actions of mkosi are also stable, but still unfortunately low coverage level, and lacks some functionalities, e.g. #35277. Killing some tests on CentOS CIs in such situation is not acceptable. Strong NAK.

I cannot follow why you want to kill whole tests rather than adding single line condition for the new test.

> It is meaningless, as half the things are silently skipped.

That contradicts what you said recently somewhere... Even half of them are skipped, running rest of the half has some meaning.

> So once again, if you'd like TEST-74 to run in nspawn, please send a follow-up PR later to move the tests that require VM isolation out of it, and into another test that is marked for VM, and then it can be moved back to not requiring a VM. The current setup is broken and must be fixed.

That may be an option, but that does NOT mean we should/can disable whole TEST-74 running on nspawn now.

Reply:
> The legacy test framework will though, as it's just a burden at this point.

I agree it's a burden (I switched from it a long time ago) but it still somehow powers the most useful upstream CI with ASan/UBSan, coredumps and coverage reports and it doesn't look like all those thing are on their way yet.

> It is meaningless, as half the things are silently skipped

This PR helps to silently skip things as well.

> Nobody runs the legacy shell tests locally anymore

Did SUSE switch to mkosi? I haven't noticed they mentioned it anywhere in the PR where the switch was supposed to take place.

Reply:
> > I cannot follow why you want to kill whole tests rather than adding single line condition for the new test.
> 
> Nothing is being killed.

This kills the test on nspawn on centos CIs. My point is only on that. And I cannot accept that.

Let me confirm that which CI(s) is the change for? Not sure, but nowadays most CIs use mkosi based test runner, no? And this is NOT for mkosi based test runner. So, the change is only for CentOS CIs?? Are there any other CIs that still use the 'legacy' bash script? Could you point out if I misunderstand something?

- 'legacy' bash test runner: run both on nspawn and qemu, used by CentOS CIs,
- 'modern' mkosi test runner: run only qemu, used by GitHub action, Ubuntu CIs (is it true?), Fedora testing farm (?), (and more?) 

Is this correct?

Reply:
> No it doesn't run both

Here's the code responsible for running the tests. It calls both `run_qemu` and `run_nspawn` (unless the variables are tweaked). It even runs stuff in both privileged and unprivileged containers sometimes to trigger various codepaths (because that's what the tests are for. to trigger as many codepaths as possible in various relevant environments):
https://github.com/systemd/systemd/blob/b7eefa1996f33f29b32bc9e5f20f4e70fa22a14a/test/test-functions#L3283-L3312

Here's the CentOS CI log https://jenkins-systemd.apps.ocp.cloud.ci.centos.org/job/upstream-vagrant-archlinux/19236/artifact//systemd-centos-ci/artifacts_2f66stpm/vagrant-logs.PYi/vagrant-arch-testsuite.FPl/TEST-74-AUX-UTILS_PASS.log

both qemu and nspawn can be seen there:
```sh
+ env --unset=UNIFIED_CGROUP_HIERARCHY --unset=SYSTEMD_NSPAWN_UNIFIED_HIERARCHY timeout --foreground 900 /systemd-meson-build/systemd-nspawn --register=no --kill-signal=SIGKILL --directory=/var/tmp/systemd-test-TEST-74-AUX-UTILS/root --setenv=SYSTEMD_UNIT_PATH=/usr/lib/systemd/tests/testdata/TEST-74-AUX-UTILS.units:/usr/lib/systemd/tests/testdata/units: --machine=TEST-74-AUX-UTILS --private-network /usr/lib/systemd/systemd systemd.unit=testsuite.target systemd.wants=TEST-74-AUX-UTILS.service systemd.wants=end.service
+ tee /var/tmp/systemd-test-TEST-74-AUX-UTILS/console.log
```
and
```sh
+ timeout --foreground 900 /sbin/qemu-system-x86_64 -smp 18 -net none -m 768M -nographic -kernel /boot/85a502bdbc514ac38f92deda4d4fc9e3/6.11.7-arch1-1/linux -drive format=raw,cache=unsafe,file=/var/tmp/systemd-test-TEST-74-AUX-UTILS/default.img -device virtio-rng-pci,max-bytes=1024,period=1000 -device vhost-vsock-pci,guest-cid=15551 -initrd /var/tmp/initrd-testsuite-vgt.img -machine accel=kvm -enable-kvm -cpu host -append 'root=LABEL=systemd_boot rw raid=noautodetect rd.luks=0 loglevel=2 init=/usr/lib/systemd/systemd console=ttyS0 SYSTEMD_UNIT_PATH=/usr/lib/systemd/tests/testdata/TEST-74-AUX-UTILS.units:/usr/lib/systemd/tests/testdata/units: systemd.unit=testsuite.target systemd.wants=TEST-74-AUX-UTILS.service noresume oops=panic panic=1 softlockup_panic=1 systemd.wants=end.service '
+ tee /var/tmp/systemd-test-TEST-74-AUX-UTILS/console.log
```